third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/CodePartitionUtility.cpp

```diff
@@ -0,0 +1,181 @@
+#include "CodePartitionUtility.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+#include "nvidia/hopper/include/Transforms/Passes.h"
+#include <list>
+#include <unordered_set>
+
+namespace tt = mlir::triton;
+namespace ttg = mlir::triton::gpu;
+namespace ttng = ::mlir::triton::nvidia_gpu;
+namespace mlir {
+
+#define DEBUG_TYPE "nvgpu-ws-utility"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+// Check to see if op is enclosed under ifOp.
+bool enclosing(scf::IfOp ifOp, Operation *op) {
+  auto pOp = op->getParentOfType<scf::IfOp>();
+  while (pOp) {
+    if (pOp == ifOp)
+      return true;
+    pOp = pOp->getParentOfType<scf::IfOp>();
+  }
+  return false;
+}
+
+bool enclosing(scf::ForOp forOp, Operation *op) {
+  auto pOp = op->getParentOfType<scf::ForOp>();
+  while (pOp) {
+    if (pOp == forOp)
+      return true;
+    pOp = pOp->getParentOfType<scf::ForOp>();
+  }
+  return false;
+}
+
+// Check to see if there is no outer loop that is enclosed under ifOp.
+bool immediateEnclosing(scf::IfOp ifOp, Operation *subOp) {
+  auto pOp = subOp->getParentOfType<scf::ForOp>();
+  if (!pOp)
+    return true;
+  return !enclosing(ifOp, pOp.getOperation());
+}
+
+// Return number of AccumCnts for the given ctrlOp. We need one for each nested
+// region that contains a channel.
+unsigned getAccumCnts(Operation *ctrlOp,
+                      const DenseSet<Operation *> &regionsWithChannels) {
+  unsigned cnt = 0;
+  LDBG("getAccumCnts: " << ctrlOp);
+  for (auto *op : regionsWithChannels) {
+    LDBG("-- getAccumCnts: " << ctrlOp << " regionsWithChannels " << op);
+    if (ctrlOp == op) {
+      ++cnt;
+      continue;
+    }
+    if (auto forOp = dyn_cast<scf::ForOp>(ctrlOp))
+      if (enclosing(forOp, op))
+        ++cnt;
+    if (auto ifOp = dyn_cast<scf::IfOp>(ctrlOp))
+      if (enclosing(ifOp, op))
+        ++cnt;
```

Comment:
this should crash or assert if an unknown region operation was encountered, such as `scf.while`